File: libro-redes-sociales-2026/scripts/compara_redes.R
# compara_redes.R — Compara tres modelos clásicos de red
# Uso (desde la carpeta raíz del repositorio):
#   Rscript libro-redes-sociales-2026/scripts/compara_redes.R

suppressPackageStartupMessages(library(igraph))
set.seed(2026)

n <- 200   # nodos
m <- 600   # aristas (aprox. las mismas en los tres modelos)

redes <- list(
  aleatoria     = sample_gnm(n, m),                      # Erdős–Rényi G(n,m)
  mundo_pequeno = sample_smallworld(1, n, 3, p = 0.05),  # Watts–Strogatz
  libre_escala  = sample_pa(n, m = 3, directed = FALSE)  # Barabási–Albert
)

resumen <- data.frame(
  red           = names(redes),
  nodos         = sapply(redes, vcount),
  aristas       = sapply(redes, ecount),
  densidad      = round(sapply(redes, edge_density), 4),
  transitividad = round(sapply(redes, transitivity), 4),
  dist_media    = round(sapply(redes, mean_distance), 2),
  grado_max     = sapply(redes, function(g) max(degree(g)))
)

print(resumen, row.names = FALSE)

salida <- "libro-redes-sociales-2026/outputs/comparacion_redes_R.csv"
write.csv(resumen, salida, row.names = FALSE)
cat("\nGuardado en", salida, "\n")

File: libro-redes-sociales-2026/scripts/compara_redes.py
# compara_redes.py — Compara tres modelos clásicos de red
# Uso (desde la carpeta raíz del repositorio):
#   python3 libro-redes-sociales-2026/scripts/compara_redes.py
# Requiere: pip install networkx

import csv

import networkx as nx

n, m = 200, 600  # nodos y aristas (aprox. las mismas en los tres modelos)

redes = {
    "aleatoria": nx.gnm_random_graph(n, m, seed=2026),            # Erdős–Rényi G(n,m)
    "mundo_pequeno": nx.watts_strogatz_graph(n, 6, 0.05, seed=2026),  # Watts–Strogatz
    "libre_escala": nx.barabasi_albert_graph(n, 3, seed=2026),    # Barabási–Albert
}

filas = []
for nombre, G in redes.items():
    filas.append({
        "red": nombre,
        "nodos": G.number_of_nodes(),
        "aristas": G.number_of_edges(),
        "densidad": round(nx.density(G), 4),
        "transitividad": round(nx.transitivity(G), 4),
        "dist_media": round(nx.average_shortest_path_length(G), 2),
        "grado_max": max(d for _, d in G.degree()),
    })

for fila in filas:
    print(fila)

salida = "libro-redes-sociales-2026/outputs/comparacion_redes_python.csv"
with open(salida, "w", newline="") as f:
    w = csv.DictWriter(f, fieldnames=filas[0].keys())
    w.writeheader()
    w.writerows(filas)

print(f"\nGuardado en {salida}")
